maintenance/githubutil/githubutil.go

```diff
@@ -0,0 +1,195 @@
+			c.limitRate(&rlErr.Rate)
+		} else if tfaErr, ok := err.(*github.TwoFactorAuthError); ok {
+			return tfaErr
+		}
+
+		if retryCount == c.retries {
+			return err
+		} else {
+			glog.Errorf("error %s: %v. Will retry.\n", action, err)
+			c.retrySleep(retryCount)
+		}
+	}
+	return err
+}
+
+// CreateStatus creates or updates a status context on the indicated reference.
+// This function limits rate and does retries if needed.
+func (c *Client) CreateStatus(owner, repo, ref string, status *github.RepoStatus) (*github.RepoStatus, *github.Response, error) {
+	glog.Infof("CreateStatus for ref '%s': %s:%s (%s)\n", ref, *status.Context, *status.State, *status.Description)
+	if c.dryRun {
+		return nil, nil, nil
+	}
+	var result *github.RepoStatus
+	var resp *github.Response
+	call := func() (*github.Response, error) {
+		var err error
+		result, resp, err = c.repoService.CreateStatus(owner, repo, ref, status)
+		return resp, err
+	}
+	err := c.withRetryAndRateLimit(call, fmt.Sprintf("creating status for ref '%s'", ref))
+	return result, resp, err
+}
+
+// GetCombinedStatus retrieves the CominedStatus for the specified reference.
+// This function limits rate, does retries if needed and handles pagination.
+func (c *Client) GetCombinedStatus(owner, repo, ref string) (*github.CombinedStatus, *github.Response, error) {
+	var status, result *github.CombinedStatus
+	var resp *github.Response
+
+	listOpts := &github.ListOptions{Page: 1, PerPage: 100}
+	lastPage := 1
+
+	call := func() (*github.Response, error) {
+		var err error
+		status, resp, err = c.repoService.GetCombinedStatus(owner, repo, ref, listOpts)
+		return resp, err
+	}
+	action := fmt.Sprintf("getting combined status for ref '%s'", ref)
+
+	for ; listOpts.Page <= lastPage; listOpts.Page++ {
+		if err := c.withRetryAndRateLimit(call, action); err != nil {
+			return result, resp, err
+		}
+		lastPage = resp.LastPage
+		if result == nil {
+			result = status
+		} else {
+			result.Statuses = append(result.Statuses, status.Statuses...)
+		}
+	}
+	return result, resp, nil
+}
+
+type PRMungeFunc func(*github.PullRequest) error
+
+// ForEachPR iterates over all PRs that fit the specified criteria, calling the munge function on every PR.
+// This function limits rate, does retries if needed, and handles pagination.
+func (c *Client) ForEachPR(owner, repo string, opts *github.PullRequestListOptions, mungePR PRMungeFunc) error {
+	var list []*github.PullRequest
+	var resp *github.Response
+
+	opts.ListOptions.Page = 1
+	opts.ListOptions.PerPage = 100
+	lastPage := 1
+
+	call := func() (*github.Response, error) {
+		var err error
+		list, resp, err = c.prService.List(owner, repo, opts)
+		return resp, err
+	}
+
+	for ; opts.ListOptions.Page <= lastPage; opts.ListOptions.Page++ {
+		if err := c.withRetryAndRateLimit(call, "processing PRs"); err != nil {
+			return err
+		}
+		lastPage = resp.LastPage
+		for _, pr := range list {
+			if err := mungePR(pr); err != nil {
```

Comment:
I'm stopping on the first error so that the error can be addressed instead of blindly continuing to the remaining PRs which may also be subject to the same error. That being said it may be useful to continue on error if there are a few PRs that are causing problems or something like that. I added another flag that makes this configurable, but defaults to failing on the first error. I added a comment to migratestatus.go to explain the choice of default.